packages/in_app_purchase/ios/Classes/InAppPurchasePlugin.m

```diff
@@ -70,11 +123,150 @@ - (void)handleProductRequestMethodCall:(FlutterMethodCall *)call result:(Flutter
+    result([FlutterError errorWithCode:@"storekit_invalide_argument"
+                               message:@"Argument type of addPayment is not a map"
+                               details:call.arguments]);
+    return;
+  }
+  NSDictionary *paymentMap = (NSDictionary *)call.arguments;
+  NSString *productID = [paymentMap objectForKey:@"productID"];
+  SKPayment *payment = [self.paymentsCache objectForKey:productID];
+  // User can use  payment object with mutable = true and add simulatesAskToBuyInSandBox = true to
+  // test the payment flow.
+  if (!payment || [paymentMap[@"mutable"] boolValue] == YES) {
+    SKMutablePayment *mutablePayment = [[SKMutablePayment alloc] init];
+    mutablePayment.productIdentifier = productID;
+    NSNumber *quantity = [paymentMap objectForKey:@"quantity"];
+    if (quantity) {
+      mutablePayment.quantity = quantity.integerValue;
+    }
+    NSString *applicationUsername = [paymentMap objectForKey:@"applicationUsername"];
+    mutablePayment.applicationUsername = applicationUsername;
+    if (@available(iOS 8.3, *)) {
+      mutablePayment.simulatesAskToBuyInSandbox =
+          [[paymentMap objectForKey:@"simulatesAskToBuyInSandBox"] boolValue];
+    }
+    [self.paymentQueueHandler addPayment:payment];
+  } else {
+    [self.paymentQueueHandler addPayment:payment];
+  }
+}
+
+- (void)finishTransaction:(FlutterMethodCall *)call result:(FlutterResult)result {
+  if (![call.arguments isKindOfClass:[NSString class]]) {
+    result([FlutterError errorWithCode:@"storekit_invalide_argument"
+                               message:@"Argument type of finishTransaction is not a string."
+                               details:call.arguments]);
+    return;
+  }
+  NSString *identifier = call.arguments;
+  SKPaymentTransaction *transaction =
+      [self.paymentQueueHandler.transactions objectForKey:identifier];
+  if (!transaction) {
+    result([FlutterError errorWithCode:@"storekit_platform_invalid_transaction"
+                               message:@"Invalid transaction ID is used."
+                               details:call.arguments]);
+    return;
+  }
+  @try {
+    // finish transaction will throw exception if the transaction type is purchasing. Notify dart
+    // about this exception.
+    [self.paymentQueueHandler finishTransaction:transaction];
+  } @catch (NSException *e) {
+    result([FlutterError errorWithCode:@"storekit_finish_transaction_exception"
+                               message:e.name
+                               details:e.description]);
+    return;
+  }
+}
+
+#pragma mark - delegates
+
+- (void)handleTransactionsUpdated:(NSArray<SKPaymentTransaction *> *)transactions {
+  NSMutableArray *maps = [NSMutableArray new];
+  for (SKPaymentTransaction *transcation in transactions) {
+    [maps addObject:[FIAObjectTranslator getMapFromSKPaymentTransaction:transcation]];
+  }
+  [self.callbackChannel invokeMethod:@"updatedTransaction" arguments:maps];
```

Comment:
nit: I think it would be a little clearer to name these methods "<callbacktype>Transactions" here and below, since there's multiple transactions in each method.